Expose the stopped property from the CassandraDatacenter spec [K8SSAND-1168]

adutra opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with main
-->

**What this PR does**:
Expose the stopped property from the CassandraDatacenter spec.

**Which issue(s) this PR fixes**:
Fixes #296 

**Checklist**
- [ ] Changes manually tested
- [ ] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CHANGELOG.md updated (not required for documentation PRs)
- [ ] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

adutra commented on controllers/k8ssandra/datacenters.go:
```diff
@@ -126,23 +126,39 @@ func (r *K8ssandraClusterReconciler) reconcileDatacenters(ctx context.Context, k
 				}
+
+				if annotations.HasAnnotationWithValue(kc, api.RebuildDcAnnotation, dcKey.Name) {
+					if recResult := r.reconcileDcRebuild(ctx, kc, actualDc, remoteClient, logger); recResult.Completed() {
```
@jsanda if a dc is added in stopped state, should we call `reconcileDcRebuild` for it?

burmanm replied:
We should throw a validation error in that case. cass-operator will not finish building the new DC if it's in stopped state.

adutra replied:
Hm that's crazy indeed. I just created a new cassdc in stopped state, and cass-operator happily created the service and even the empty statefulsets for me. No validation errors. Conditions Initialized and Stopped are true. This is not going to work, I will add a validation error to forbid creating a dc in stopped state.

adutra commented on controllers/k8ssandra/stop_dc_test.go:
```diff
@@ -0,0 +1,362 @@
+// stopDc tests scenarios that involve stopping a CassandraDatacenter.
+func stopDc(t *testing.T, ctx context.Context, f *framework.Framework, _ string) {
+	t.Run("StopExistingDatacenter", stopDcTest(f, ctx, stopExistingDc))
+	t.Run("AddAndStopDatacenter", stopDcTest(f, ctx, addAndStopDc))
```
I'm not sure this test is useful, I might remove it. There is nothing particular about stopping a dc that wasn't part of the initial cluster.

adutra commented on controllers/k8ssandra/datacenters.go:
```diff
@@ -125,23 +125,39 @@ func (r *K8ssandraClusterReconciler) reconcileDatacenters(ctx context.Context, k
 
-			if !cassandra.DatacenterReady(actualDc) {
-				logger.Info("Waiting for datacenter to become ready")
+			if actualDc.Status.CassandraOperatorProgress != cassdcapi.ProgressReady {
```
But we were already doing this. Prior to this change we would call `DatacenterReady` here. It is implemented as follows:

```go
func DatacenterReady(dc *cassdcapi.CassandraDatacenter) bool {
	return dc.GetConditionStatus(cassdcapi.DatacenterReady) == corev1.ConditionTrue && dc.Status.CassandraOperatorProgress == cassdcapi.ProgressReady
}
```

The only reason why I extracted this comparison out of `DatacenterReady` is because there are two functions, `DatacenterStopped` and `DatacenterStopping` that differ only by the handling of `CassandraOperatorProgress`. 

Do you advocate instead for changing how `DatacenterReady` is implemented? I wanted to avoid this.

jsanda replied:
I might be in favor of changing `DatacenterReady`. It doesn't have to be done in this PR (if it is done at all).

adutra replied:
OK I removed the if block and came back to something closer to what we have in main. We can then see later to change DatacenterReady, but it is called in many places, so definitely something we want to do in another PR.

adutra commented on controllers/k8ssandra/datacenters.go:
```diff
@@ -125,23 +125,39 @@ func (r *K8ssandraClusterReconciler) reconcileDatacenters(ctx context.Context, k
-				logger.Info("Waiting for datacenter to become ready")
+			if actualDc.Status.CassandraOperatorProgress != cassdcapi.ProgressReady {
+				logger.Info("Waiting for datacenter to finish updating")
 				return result.RequeueSoon(r.DefaultDelay), actualDcs
```
OK I will replace `result.RequeueSoon(r.DefaultDelay)` with `result.Done()` – even if this wasn't changed by the current PR.

adutra commented on controllers/k8ssandra/datacenters.go:
```diff
@@ -125,27 +131,47 @@ func (r *K8ssandraClusterReconciler) reconcileDatacenters(ctx context.Context, k
+			if actualDc.Spec.Stopped {
+				if !cassandra.DatacenterStopped(actualDc) {
+					logger.Info("Waiting for datacenter to satisfy Stopped condition")
+					return result.RequeueSoon(r.DefaultDelay), actualDcs
```
> And if the datacenter is stopped, can't we just move on to the next one?

Here we are waiting for the Stopped condition to be true, since that's what we want. When it becomes true, we would move to the next dc.

adutra commented on controllers/k8ssandra/schemas.go:
```diff
@@ -62,6 +62,14 @@ func (r *K8ssandraClusterReconciler) checkSchemaAgreement(mgmtApi cassandra.Mana
+	for uid := range versions {
+		// a key named UNREACHABLE may appear in the results when nodes are unreachable.
+		// We exclude these keys from the check.
+		if uid == "UNREACHABLE" {
```
I had to introduce this to allow the operator to make progress when a dc is stopped.

When any node is unreachable, the management api returns a map that looks like below:

```json
{
  "UNREACHABLE": [
    "172.18.0.2"
  ],
  "aa573028-7c70-3b8f-a247-13bbca2011d2": [
    "172.18.0.9",
    "172.18.0.6"
  ]
}
```

If you don't exclude the UNREACHABLE entry, this method will always return an error and the operator will get stuck.

Note: the output of `nodetool describecluster` is similar, so it's not a problem with the management API.

adutra commented on test/e2e/stargate_apis_test.go:
```diff
@@ -256,7 +265,7 @@ func createKeyspaceAndTableNative(t *testing.T, connection *client.CqlClientConn
 	require.IsType(t, &message.SchemaChangeResult{}, response.Body.Message, "Expected CREATE TABLE response to be of type SchemaChangeResult")
 	response = sendQuery(t, connection, fmt.Sprintf("TRUNCATE %s.%s", keyspaceName, tableName))
-	assert.IsType(t, &message.VoidResult{}, response.Body.Message, "Expected TRUNCATE response to be of type VoidResult")
+	require.IsType(t, &message.VoidResult{}, response.Body.Message, "Expected TRUNCATE response to be of type VoidResult")
```
Just fixing a few panics I've seen while working on this PR.

jsanda commented on controllers/k8ssandra/k8ssandracluster_controller.go:
```diff
@@ -160,7 +160,7 @@ func (r *K8ssandraClusterReconciler) afterCassandraReconciled(ctx context.Contex
 			return recResult
 		} else if recResult := r.reconcileStargate(ctx, kc, dcTemplate, dc, logger, remoteClient); recResult.Completed() {
```
I am inclined to say we do nothing. There won't be any pods matching the selector so I would assume Prometheus won't try to scrape anything.

Miles-Garnsey replied:
Sorry, I didn't spot this earlier. @jsanda is right, nothing will be scraped. 

From what I can understand `stopped` is intended as a temporary state and we expect the DC to eventually start again. On that basis, I agree with John in saying that nothing should happen with telemetry.

jsanda reviewed (changes requested):
I did some manual testing and it reminded me of some additional changed that I forgot to include in the issue. When I stop a DC, it gets removed from the replication strategy for all keyspaces. We shouldn't be removing replicas.

`reconcileStargateAuthSchema` and `reconcileReaperSchema` both call `kc.GetReadyDatacenters` followed by `cassandra.ComputeReplicationFromDcTemplates`. Stopped DCs should be included here when computing replication. 

Similarly for system keyspaces, `cassandra.GetDatacentersForSystemReplication` should be updated to include stopped as well as ready DCs.

We do update replication of user-defined keyspaces during a rebuild. I checked that code. I don't think it needs any changes because it only checked if the `Initialized` condition is set on the CassandraDatacenter.

Lastly, tests should be updated to make sure replication changes are not made.

adutra commented:
> When I stop a DC, it gets removed from the replication strategy for all keyspaces. We shouldn't be removing replicas.

By not removing the stopped dc from replication strategies, some CLs like QUORUM would start failing. Is that OK in your opinion?

> Similarly for system keyspaces, cassandra.GetDatacentersForSystemReplication should be updated to include stopped as well as ready DCs.

👍 

> reconcileStargateAuthSchema and reconcileReaperSchema both call kc.GetReadyDatacenters followed by cassandra.ComputeReplicationFromDcTemplates. Stopped DCs should be included here when computing replication.

In this case is it enough to replace `kc.GetReadyDatacenters` with `cassandra.GetDatacentersForSystemReplication`? It seems it's going to be the same algorithm.

adutra commented on controllers/k8ssandra/stargate.go:
```diff
@@ -36,6 +36,11 @@ func (r *K8ssandraClusterReconciler) reconcileStargate(
 	actualStargate := &stargateapi.Stargate{}
 	logger = logger.WithValues("Stargate", stargateKey)
 
+	if actualDc.Spec.Stopped && stargateTemplate != nil {
```
I've asked myself this question too. I believe a Stargate node can operate normally even if the local DC is down. It could be a good thing to keep it up and running.

adutra commented:
@jsanda I believe I addressed all your comments.

Regarding `GetDatacentersForSystemReplication` I believe it can be used for all keyspaces, in all situations. I updated the function description. Let me know if that makes sense or if I've gone completely nuts.

jsanda commented on pkg/cassandra/util.go:
```diff
@@ -33,35 +33,39 @@ func DatacenterStopping(dc *cassdcapi.CassandraDatacenter) bool {
+// through CQL, and we cannot include non-existing DCs anymore; this is going to be the case: 1) when updating system
+// keyspaces; 2) when configuring keyspaces for Stargate and Reaper; and 3) when updating user keyspaces during a DC
+// rebuild.
+func GetDatacentersForReplication(kc *api.K8ssandraCluster) []api.CassandraDatacenterTemplate {
```
I don't think combining the functions is going to work. We do the schema checks after reconciling each DC. Let's say we are creating a cluster with 3 DCs - dc1, dc2, dc3. The `CassandraInitialized` condition won't get set until all three DCs are ready, but the schema checks happen after we finish reconciling each DC (i.e., each DC is ready or stopped). With this change we will include non-existing DCs.

I did a quick test to verify. I created this cluster:

```yaml
apiVersion: k8ssandra.io/v1alpha1
kind: K8ssandraCluster
metadata:
  name: test
spec:
  cassandra:
    serverVersion: "4.0.1"
    storageConfig:
      cassandraDataVolumeClaimSpec:
        storageClassName: standard
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 5Gi
    config:
      cassandraYaml:
        auto_snapshot: false
        memtable_flush_writers: 1
        commitlog_segment_size_in_mb: 2
        concurrent_compactors: 1
        compaction_throughput_mb_per_sec: 0
        sstable_preemptive_open_interval_in_mb: 0
        key_cache_size_in_mb: 0
        thrift_prepared_statements_cache_size_mb: 1
        prepared_statements_cache_size_mb: 1
        slow_query_log_timeout_in_ms: 0
        counter_cache_size_in_mb: 0
        concurrent_reads: 2
        concurrent_writes: 2
        concurrent_counter_writes: 2
      jvmOptions:
        heapSize: 512Mi
    networking:
      hostNetwork: true
    resources:
      limits:
        memory: 1024Mi
    datacenters:
      - metadata:
          name: dc1
        k8sContext: kind-k8ssandra-0
        size: 1
      - metadata:
          name: dc2
        k8sContext: kind-k8ssandra-1
        size: 1
  stargate:
    size: 1
    heapSize: 512Mi
    resources:
      limits:
        memory: 1024Mi
  reaper:
    autoScheduling:
      enabled: true
```
And as expected the schema checks for Stargate fail. Here is the error from Cassandra's logs:

```
INFO  [epollEventLoopGroup-7-3] 2022-02-04 03:52:22,178 RpcExecutionException.java:39 - Failed to execute method NodeOps.createKeyspace
org.apache.cassandra.exceptions.ConfigurationException: Unrecognized strategy option {dc2} passed to NetworkTopologyStrategy for keyspace data_endpoint_auth
```

adutra commented:
Putting this back to draft. It's becoming way more complex than I initially thought. I might give up on finishing this by the way.

Also the latest rebase broke the e2e test. Reaper is not able anymore to start with one DC dead:

```
INFO   [2022-02-04 21:19:32,966] [main] c.d.d.c.Cluster - New Cassandra host cluster1-dc2-service/172.18.0.7:9042 added
com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (tried: cluster1-dc2-service/172.18.0.7:9042 (com.datastax.driver.core.exceptions.UnavailableException: Not enough replicas available for query at consistency QUORUM (2 required but only 1 alive)))
	at com.datastax.driver.core.exceptions.NoHostAvailableException.copy(NoHostAvailableException.java:83)
	at com.datastax.driver.core.exceptions.NoHostAvailableException.copy(NoHostAvailableException.java:37)
	at com.datastax.driver.core.DriverThrowables.propagateCause(DriverThrowables.java:35)
	at com.datastax.driver.core.DefaultResultSetFuture.getUninterruptibly(DefaultResultSetFuture.java:293)
	at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:58)
	at org.cognitor.cassandra.migration.Database.getVersion(Database.java:195)
	at io.cassandrareaper.storage.CassandraStorage.initializeCassandraSchema(CassandraStorage.java:275)
	at io.cassandrareaper.storage.CassandraStorage.initializeAndUpgradeSchema(CassandraStorage.java:257)
	at io.cassandrareaper.storage.CassandraStorage.<init>(CassandraStorage.java:245)
	at io.cassandrareaper.storage.InitializeStorage.initializeStorageBackend(InitializeStorage.java:65)
	at io.cassandrareaper.ReaperDbMigrationCommand.run(ReaperDbMigrationCommand.java:50)
	at io.cassandrareaper.ReaperDbMigrationCommand.run(ReaperDbMigrationCommand.java:28)
	at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:98)
	at io.dropwizard.cli.Cli.run(Cli.java:78)
	at io.dropwizard.Application.run(Application.java:94)
	at io.cassandrareaper.ReaperApplication.main(ReaperApplication.java:102)
```

I will investigate on Monday.

adutra commented:
I take it back: it might be a quick win to just alter Reaper's keyspace. I will try that route first.

adutra commented on pkg/reaper/resource.go:
```diff
@@ -64,6 +65,12 @@ func NewReaper(
+	// If the cluster is already initialized and some DCs are flagged as stopped, we cannot achieve QUORUM in the
+	// cluster for Reaper's keyspace. In this case we simply skip schema migration, otherwise Reaper wouldn't be able to
+	// start up.
+	if kc.Status.GetConditionStatus(k8ssandraapi.CassandraInitialized) == corev1.ConditionTrue && kc.HasStoppedDatacenters() {
```
I'm hesitant here. What conditions should we check to decide to skip schema migration?

jsanda replied:
I deployed a cluster with Reaper `deploymentMode: SINGLE`. Reaper is deployed to dc1. I go back and forth stopping/starting dc2. It causes an update of the Reaper Deployment. That should be avoidable. I'm ok with fixing it in a follow up issue if you prefer since it is arguably an optimization.

adutra commented on test/e2e/stargate_apis_test.go:
```diff
@@ -255,8 +255,6 @@ func createKeyspaceAndTableNative(t *testing.T, connection *client.CqlClientConn
 	))
 	require.IsType(t, &message.SchemaChangeResult{}, response.Body.Message, "Expected CREATE TABLE response to be of type SchemaChangeResult")
-	response = sendQuery(t, connection, fmt.Sprintf("TRUNCATE %s.%s", keyspaceName, tableName))
-	assert.IsType(t, &message.VoidResult{}, response.Body.Message, "Expected TRUNCATE response to be of type VoidResult")
```
This TRUNCATE query fails randomly with TruncateError. I think we can safely remove it.

adutra commented:
This is again more or less ready for another review. I'm not entirely happy with the workaround for the Reaper issue, but the test is passing.

jsanda reviewed (approved):
I left a few comments in my last review, but I think they can be addressed in subsequent issues.
